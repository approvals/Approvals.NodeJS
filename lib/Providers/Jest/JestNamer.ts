import { expect } from "@jest/globals";
import path from "path";
const Namer:any = require( "../../Namer");

export function convertToFilename(name: string):string {
    const forbidden = "\\/:?\"<>|' ";
    for (let forbiddenKey of forbidden) {
        if (name.includes(forbiddenKey)) {
            name = name.replace(forbiddenKey, "_");
        }
    }
    return name;
}

export function getJestNamer(): typeof Namer{
    const state = expect.getState();
    const file = path.parse(state.testPath as string);
    const testPath = file.dir;
    const testFileName = file.name;
    const testName = convertToFilename(`${testFileName}.${state.currentTestName}`);
    return new Namer(testPath, testName);
}
