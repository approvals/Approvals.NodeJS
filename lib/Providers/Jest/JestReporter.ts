import {expect} from "@jest/globals";
import type {ApprovalFailureReporter} from "../../Core/ApprovalFailureReporter";

const fs = require('fs');

export class JestReporter implements ApprovalFailureReporter {
    name: string = "JestReporter";

    canReportOn() {
        return true;
    }

    report(approvedFilePath: string, receivedFilePath: string) {
        const approvedText = fs.readFileSync(approvedFilePath).toString();
        const receivedText = fs.readFileSync(receivedFilePath).toString();
        expect(receivedText).toBe(approvedText);
    }
}
