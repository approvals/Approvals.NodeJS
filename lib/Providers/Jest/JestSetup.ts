import { ApprovedFileLog } from "../../Logs/ApprovedFileLog";
import { FailedFileLog } from "../../Logs/FailedFileLog";
import fs from "fs";
import { LogUtils } from "../../Logs/LogUtils";

export let jestInitializingRun = false;

export function initializeGlobalsForJest() {
  ApprovedFileLog.forceClearLogFile();
  FailedFileLog.forceClearFileLog();
  jestInitializingRun = true;
}

export function helpUserSetupJest(fileToCheck: string) {
  LogUtils.ensureTempDirectoryExists();
  // Check if the file exists
  if (fs.existsSync(fileToCheck)) {
    // Get the date that fileToCheck was created
    const fileStats = fs.statSync(fileToCheck);
    const fileCreationDate = fileStats.birthtime;

    // Check if the date was in the last 24 hours
    const currentDate = new Date();
    const twentyFourHoursAgo = new Date(
      currentDate.getTime() - 24 * 60 * 60 * 1000,
    );
    const isFileRecent = fileCreationDate > twentyFourHoursAgo;

    if (isFileRecent) {
      return;
    }
  }
  const helpMessage = `
**********************************************
Jest needs to be set up for logging.
More information at https://github.com/approvals/Approvals.NodeJS/blob/master/docs/how_tos/InitializeJestForLogging.md
**********************************************
`;
  console.log(helpMessage);
}
