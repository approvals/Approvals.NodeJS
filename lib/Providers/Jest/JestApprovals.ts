const StringWriter = require("../../StringWriter");
const approvals = require("../../Approvals");
import { Options } from "../../Core/Options";
import { getJestNamer } from "./JestNamer";


export function verify(sut: any, options?: Options): void {
    const config = approvals.getConfig();
    config.reporters = ["diffmerge"];
    options = options || new Options()
    const writer = new StringWriter(config,  `${sut}`, options.forFile().getFileExtension());
    approvals.verifyWithControl(getJestNamer(), writer, null, config);
}

export function verifyAsJson(data: any, options?: Options): void {
    const text = JSON.stringify(data, null, "  ");
    options = options || new Options()
    options = options.forFile().withFileExtention(".json")
    verify(text, options);
}
