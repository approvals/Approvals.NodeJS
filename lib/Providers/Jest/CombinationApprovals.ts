import {verify} from "./JestApprovals";
import {printCombinations, EMPTY} from  "../../Utilities/Printers"

export function verifyAllCombinations1<T1>(func: (i: T1) => any, params1: T1[]) {
    // @ts-ignore
    verify(printCombinations((t1, t2, t3, t4, t5, t6, t7, t8, t9) => func(t1), params1, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}

export function verifyAllCombinations2<T1, T2>(func: (t1: T1, t2: T2) => any, params1: T1[], params2: T2[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: any, t4: any, t5: any, t6: any, t7: any, t8: any, t9: any) => func(t1, t2), params1, params2, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}


export function verifyAllCombinations3<T1, T2, T3>(func: (t1: T1, t2: T2, t3: T3) => any, params1: T1[], params2: T2[], params3: T3[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: any, t5: any, t6: any, t7: any, t8: any, t9: any) => func(t1, t2, t3), params1, params2, params3, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}

export function verifyAllCombinations4<T1, T2, T3, T4>(func: (t1: T1, t2: T2, t3: T3, t4: T4) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: any, t6: any, t7: any, t8: any, t9: any) => func(t1, t2, t3, t4), params1, params2, params3, params4, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}

export function verifyAllCombinations5<T1, T2, T3, T4, T5>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: any, t7: any, t8: any, t9: any) => func(t1, t2, t3, t4, t5), params1, params2, params3, params4, params5, EMPTY, EMPTY, EMPTY, EMPTY));
}

export function verifyAllCombinations6<T1, T2, T3, T4, T5, T6>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[], params6: T6[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: any, t8: any, t9: any) => func(t1, t2, t3, t4, t5, t6), params1, params2, params3, params4, params5, params6, EMPTY, EMPTY, EMPTY));
}

export function verifyAllCombinations7<T1, T2, T3, T4, T5, T6,T7>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7:T7) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[], params6: T6[], params7: T7[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: T7, t8: any, t9: any) => func(t1, t2, t3, t4, t5, t6,t7), params1, params2, params3, params4, params5, params6, params7, EMPTY, EMPTY));
}
export function verifyAllCombinations8<T1, T2, T3, T4, T5, T6,T7,T8>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7:T7, t8:T8) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[], params6: T6[], params7: T7[], params8: T8[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: T7, t8: T8, t9: any) => func(t1, t2, t3, t4, t5, t6,t7, t8), params1, params2, params3, params4, params5, params6, params7, params8, EMPTY));
}
export function verifyAllCombinations9<T1, T2, T3, T4, T5, T6,T7,T8,T9>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7:T7, t8:T8, t9:T9) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[], params6: T6[], params7: T7[], params8: T8[], params9: T9[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: T7, t8: T8, t9: T9) => func(t1, t2, t3, t4, t5, t6,t7, t8,t9), params1, params2, params3, params4, params5, params6, params7, params8, params9));
}
