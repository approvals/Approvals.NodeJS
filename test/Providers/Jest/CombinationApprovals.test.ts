import {describe, test} from "@jest/globals";
import {verify} from "../../../lib/Providers/Jest/JestApprovals";

const EMPTY_ENTRY = {};
const EMPTY = [EMPTY_ENTRY];

function printCombinations<T1, T2, T3, T4, T5, T6, T7, T8, T9>(func: <T1, T2, T3, T4, T5, T6, T7, T8, T9>(t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: T7, t8: T8, t9: T9) => any,
                                                               params1: T1[],
                                                               params2: T2[],
                                                               params3: T3[],
                                                               params4: T4[],
                                                               params5: T5[],
                                                               params6: T6[],
                                                               params7: T7[],
                                                               params8: T8[],
                                                               params9: T9[]
): string {
    let text = "";
    for (let p1 of params1) {
        for (let p2 of params2) {
            for (let p3 of params3) {
                for (let p4 of params4) {
                    for (let p5 of params5) {
                        for (let p6 of params6) {
                            for (let p7 of params7) {
                                for (let p8 of params8) {
                                    for (let p9 of params9) {
                                        let output;
                                        try {
                                            output = func(p1, p2, p3, p4, p5, p6, p7, p8, p9);
                                        } catch (e) {
                                            output = `${e}`;
                                        }
                                        const parameters = [p1, p2, p3, p4, p5, p6, p7, p8, p9].filter(p => p !== EMPTY_ENTRY);

                                        text += `[${parameters}] => ${output}\n`;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    return text;
}

function verifyAllCombinations1<T1>(func: (i: T1) => any, params1: T1[]) {
    // @ts-ignore
    verify(printCombinations((t1, t2, t3, t4, t5, t6, t7, t8, t9) => func(t1), params1, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}

function verifyAllCombinations2<T1, T2>(func: (t1: T1, t2: T2) => any, params1: T1[], params2: T2[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: any, t4: any, t5: any, t6: any, t7: any, t8: any, t9: any) => func(t1, t2), params1, params2, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}


function verifyAllCombinations3<T1, T2, T3>(func: (t1: T1, t2: T2, t3: T3) => any, params1: T1[], params2: T2[], params3: T3[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: any, t5: any, t6: any, t7: any, t8: any, t9: any) => func(t1, t2, t3), params1, params2, params3, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}

function verifyAllCombinations4<T1, T2, T3, T4>(func: (t1: T1, t2: T2, t3: T3, t4: T4) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: any, t6: any, t7: any, t8: any, t9: any) => func(t1, t2, t3, t4), params1, params2, params3, params4, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}

function verifyAllCombinations5<T1, T2, T3, T4, T5>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: any, t7: any, t8: any, t9: any) => func(t1, t2, t3, t4, t5), params1, params2, params3, params4, params5, EMPTY, EMPTY, EMPTY, EMPTY));
}

function verifyAllCombinations6<T1, T2, T3, T4, T5, T6>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[], params6: T6[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: any, t8: any, t9: any) => func(t1, t2, t3, t4, t5, t6), params1, params2, params3, params4, params5, params6, EMPTY, EMPTY, EMPTY));
}

function verifyAllCombinations7<T1, T2, T3, T4, T5, T6,T7>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7:T6) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[], params6: T6[], params7: T7[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: T7, t8: any, t9: any) => func(t1, t2, t3, t4, t5, t6,t7), params1, params2, params3, params4, params5, params6, params7, EMPTY, EMPTY));
}


describe("Combinations", () => {
    test("verify1", () => {
        function binarySquare(i: number) {
            if (i === 2) {
                throw new Error("There is no 2");
            }
            return i;
        }

        const numbers = [0, 1, 2];
        verifyAllCombinations1(binarySquare, numbers);
    });
    test("verify2", () => {
        const numbers = [0, 1, 2];
        verifyAllCombinations2((a, b) => a + b, numbers, numbers);
    });
    test("verify3", () => {
        const numbers = [3, 4];
        verifyAllCombinations3((a, b, c) => a * b * c, numbers, numbers, numbers);
    });
    test("verify4", () => {
        const numbers = [3, 4];
        verifyAllCombinations4((a, b, c, d) => a + "" + b + c + d, numbers, numbers, numbers, numbers);
    });
    test("verify5", () => {
        verifyAllCombinations5((a, b, c, d, e) => e + d + c + b + a, ["a", "b"], ["a", "b"], ["c", "d"], ["e", "f"], ["g", "h", "i"]);
    });
    test("verify6", () => {
        verifyAllCombinations6((a, b, c, d, e, f) => "#" + a + " " + f + e + d + c + b, [1, 2], ["a", "b"], ["c", "d"], ["e", "f"], ["i", "j"], ["g", "h"]);
    });
    test("verify7", () => {
        verifyAllCombinations7((a, b, c, d, e, f, g) => a + b + c + d + e + f + g, [1, 2], [2, 3], [2, 3], [2, 3], [2, 3], [2, 3], [2, 3]);
    });
});
