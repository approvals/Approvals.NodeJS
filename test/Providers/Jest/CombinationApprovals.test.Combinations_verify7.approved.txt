[1,2,2,2,2,2,2] => 13
[1,2,2,2,2,2,3] => 14
[1,2,2,2,2,3,2] => 14
[1,2,2,2,2,3,3] => 15
[1,2,2,2,3,2,2] => 14
[1,2,2,2,3,2,3] => 15
[1,2,2,2,3,3,2] => 15
[1,2,2,2,3,3,3] => 16
[1,2,2,3,2,2,2] => 14
[1,2,2,3,2,2,3] => 15
[1,2,2,3,2,3,2] => 15
[1,2,2,3,2,3,3] => 16
[1,2,2,3,3,2,2] => 15
[1,2,2,3,3,2,3] => 16
[1,2,2,3,3,3,2] => 16
[1,2,2,3,3,3,3] => 17
[1,2,3,2,2,2,2] => 14
[1,2,3,2,2,2,3] => 15
[1,2,3,2,2,3,2] => 15
[1,2,3,2,2,3,3] => 16
[1,2,3,2,3,2,2] => 15
[1,2,3,2,3,2,3] => 16
[1,2,3,2,3,3,2] => 16
[1,2,3,2,3,3,3] => 17
[1,2,3,3,2,2,2] => 15
[1,2,3,3,2,2,3] => 16
[1,2,3,3,2,3,2] => 16
[1,2,3,3,2,3,3] => 17
[1,2,3,3,3,2,2] => 16
[1,2,3,3,3,2,3] => 17
[1,2,3,3,3,3,2] => 17
[1,2,3,3,3,3,3] => 18
[1,3,2,2,2,2,2] => 14
[1,3,2,2,2,2,3] => 15
[1,3,2,2,2,3,2] => 15
[1,3,2,2,2,3,3] => 16
[1,3,2,2,3,2,2] => 15
[1,3,2,2,3,2,3] => 16
[1,3,2,2,3,3,2] => 16
[1,3,2,2,3,3,3] => 17
[1,3,2,3,2,2,2] => 15
[1,3,2,3,2,2,3] => 16
[1,3,2,3,2,3,2] => 16
[1,3,2,3,2,3,3] => 17
[1,3,2,3,3,2,2] => 16
[1,3,2,3,3,2,3] => 17
[1,3,2,3,3,3,2] => 17
[1,3,2,3,3,3,3] => 18
[1,3,3,2,2,2,2] => 15
[1,3,3,2,2,2,3] => 16
[1,3,3,2,2,3,2] => 16
[1,3,3,2,2,3,3] => 17
[1,3,3,2,3,2,2] => 16
[1,3,3,2,3,2,3] => 17
[1,3,3,2,3,3,2] => 17
[1,3,3,2,3,3,3] => 18
[1,3,3,3,2,2,2] => 16
[1,3,3,3,2,2,3] => 17
[1,3,3,3,2,3,2] => 17
[1,3,3,3,2,3,3] => 18
[1,3,3,3,3,2,2] => 17
[1,3,3,3,3,2,3] => 18
[1,3,3,3,3,3,2] => 18
[1,3,3,3,3,3,3] => 19
[2,2,2,2,2,2,2] => 14
[2,2,2,2,2,2,3] => 15
[2,2,2,2,2,3,2] => 15
[2,2,2,2,2,3,3] => 16
[2,2,2,2,3,2,2] => 15
[2,2,2,2,3,2,3] => 16
[2,2,2,2,3,3,2] => 16
[2,2,2,2,3,3,3] => 17
[2,2,2,3,2,2,2] => 15
[2,2,2,3,2,2,3] => 16
[2,2,2,3,2,3,2] => 16
[2,2,2,3,2,3,3] => 17
[2,2,2,3,3,2,2] => 16
[2,2,2,3,3,2,3] => 17
[2,2,2,3,3,3,2] => 17
[2,2,2,3,3,3,3] => 18
[2,2,3,2,2,2,2] => 15
[2,2,3,2,2,2,3] => 16
[2,2,3,2,2,3,2] => 16
[2,2,3,2,2,3,3] => 17
[2,2,3,2,3,2,2] => 16
[2,2,3,2,3,2,3] => 17
[2,2,3,2,3,3,2] => 17
[2,2,3,2,3,3,3] => 18
[2,2,3,3,2,2,2] => 16
[2,2,3,3,2,2,3] => 17
[2,2,3,3,2,3,2] => 17
[2,2,3,3,2,3,3] => 18
[2,2,3,3,3,2,2] => 17
[2,2,3,3,3,2,3] => 18
[2,2,3,3,3,3,2] => 18
[2,2,3,3,3,3,3] => 19
[2,3,2,2,2,2,2] => 15
[2,3,2,2,2,2,3] => 16
[2,3,2,2,2,3,2] => 16
[2,3,2,2,2,3,3] => 17
[2,3,2,2,3,2,2] => 16
[2,3,2,2,3,2,3] => 17
[2,3,2,2,3,3,2] => 17
[2,3,2,2,3,3,3] => 18
[2,3,2,3,2,2,2] => 16
[2,3,2,3,2,2,3] => 17
[2,3,2,3,2,3,2] => 17
[2,3,2,3,2,3,3] => 18
[2,3,2,3,3,2,2] => 17
[2,3,2,3,3,2,3] => 18
[2,3,2,3,3,3,2] => 18
[2,3,2,3,3,3,3] => 19
[2,3,3,2,2,2,2] => 16
[2,3,3,2,2,2,3] => 17
[2,3,3,2,2,3,2] => 17
[2,3,3,2,2,3,3] => 18
[2,3,3,2,3,2,2] => 17
[2,3,3,2,3,2,3] => 18
[2,3,3,2,3,3,2] => 18
[2,3,3,2,3,3,3] => 19
[2,3,3,3,2,2,2] => 17
[2,3,3,3,2,2,3] => 18
[2,3,3,3,2,3,2] => 18
[2,3,3,3,2,3,3] => 19
[2,3,3,3,3,2,2] => 18
[2,3,3,3,3,2,3] => 19
[2,3,3,3,3,3,2] => 19
[2,3,3,3,3,3,3] => 20
